packages/cw22/src/lib.rs

```diff
@@ -0,0 +1,168 @@
+/*!
+The standard used to declare which interface contract implements
+This standard is inspired by the EIP-165 from Ethereum.
+
+For more information on this specification, please check out the
+[README](https://github.com/CosmWasm/cw-plus/blob/main/packages/cw22/README.md).
+ */
+
+mod query;
+
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::StdError;
+use cosmwasm_std::{StdResult, Storage};
+use cw_storage_plus::Map;
+use query::VersionResponse;
+use semver::{Version, VersionReq};
+
+pub const SUPPORTED_INTERFACES: Map<String, String> = Map::new("supported_interfaces");
+
+#[cw_serde]
+pub struct ContractSupportedInterface {
+    /// supported_interface is an optional parameter returning a vector of string represents interfaces
+    /// that the contract support The string value is the interface crate names in Rust crate Registry.
+    /// This parameter is inspired by the EIP-165 from Ethereum.
+    /// Each string value should follow a common standard such as <Registry Domain>:<Crate Name>
+    /// e.g "crates.io:cw2"
+    /// NOTE: this is just a hint for the caller to adapt on how to interact with this contract.
+    /// There is no guarantee that the contract actually implement these interfaces.
+    pub supported_interface: String,
+    /// semantic version on release tags.
+    /// e.g  "v0.16.0"
+    pub version: String,
+}
+
+/// set_contract_supported_interface should be used in instantiate to store the original version
+pub fn set_contract_supported_interface(
+    store: &mut dyn Storage,
+    mut supported_interfaces: Vec<ContractSupportedInterface>,
+) -> Result<(), StdError> {
+    while let Some(supported_interface) = supported_interfaces.pop() {
+        let id = supported_interface.supported_interface;
+        let version = supported_interface.version;
+        SUPPORTED_INTERFACES.save(store, id, &version)?;
+    }
+    Ok(())
+}
+
+/// query_supported_interface_version show if contract supports interface and which version
+pub fn query_supported_interface_version(
+    store: &dyn Storage,
+    interface: String,
+) -> StdResult<ContractSupportedInterface> {
+    let version = SUPPORTED_INTERFACES
+        .may_load(store, interface.clone())?
+        .unwrap_or_default();
+    let res = ContractSupportedInterface {
+        supported_interface: interface,
+        version,
+    };
+    Ok(res)
+}
+
+/// query_supported_interface show if contract supports interface with specified version
+/// query example">=1.2.3, <1.8.0"
+pub fn query_supported_interface(
+    store: &dyn Storage,
+    interface: String,
+    query: String,
+) -> StdResult<VersionResponse> {
+    let req = VersionReq::parse(&query).unwrap();
+    let supported_version_rs = SUPPORTED_INTERFACES
+        .may_load(store, interface)?
+        .unwrap_or_default();
+    let supported_version = Version::parse(&supported_version_rs);
+    match supported_version {
+        Ok(ver) => Ok(VersionResponse {
```

Comment:
since this is a query, I think just return true or false would be enough.

Reply:
it returns more information to the querier. Example which specifies version interface is supported, or not?

Reply:
I think this function can be removed.

`query_supported_interface_version` already returns the version (or missing), be DRY.
Also, we will most likely be getting this though `Map.query(...)` as we remotely query a contract (don't compare with our own contracts version). Let's make the version comparison helper separate.

I would just add two functions that don't touch storage and show the semver usage, like:

// most comparisons just require a minimum - make this easy
`minimum_version(version: &str, required: &str) -> bool {}` - maybe make `Result<(), Error>`?

// allow more powerful query strings - ">=1.2.3,<2"
`require_version(version: &str, request: &str) -> bool {}` - or Result as above

Reply:
If you go for the Result type, please define a custom Cw22Error type in the spec.

Reply:
@ethanfrey for the `minimum_version`, you mean a helper function like this?
```
pub fn minimum_version(version: &str, required: &str) -> bool {
  let req = VersionReq::parse(">={required}").unwrap();
  let version = Version::parse(version);
  req.matches(&version)
}
```